aidial_adapter_openai/utils/tokenizer.py

```diff
@@ -0,0 +1,125 @@
+        """
+        if self.model == "gpt-3.5-turbo-0301":
+            return 4
+        return 3
+
+    @property
+    def tokens_per_name(self) -> int:
+        """
+        Tokens, that are counter for "name" field in message, if it's present
+        """
+        if self.model == "gpt-3.5-turbo-0301":
+            return -1
+        return 1
+
+    def calculate_request_prompt_tokens(self, messages_tokens: int):
+        """
+        Amount of tokens, that will be counted by API
+        is greater than actual sum of tokens of all messages by PROMPT_TOKENS
+        """
+        return self.TOKENS_PER_REQUEST + messages_tokens
+
+    def calculate_prompt_tokens(self, messages: List[TokenizeInput]) -> int:
+        return self.calculate_request_prompt_tokens(
+            messages_tokens=sum(map(self.calculate_message_tokens, messages))
+        )
+
+    def available_message_tokens(self, max_prompt_tokens: int):
+        return max_prompt_tokens - self.TOKENS_PER_REQUEST
+
+    @abstractmethod
+    def calculate_message_tokens(self, message_holder: TokenizeInput) -> int:
+        pass
+
+
+class PlainTextTokenizer(BaseTokenizer[dict]):
+    """
+    Tokenizer for message.
+    Calculates only textual tokens, not image tokens.
+    """
+
+    def calculate_message_tokens(self, message_holder: dict) -> int:
+        tokens = self.tokens_per_message
+        for key, value in message_holder.items():
+            if key == "name":
+                tokens += self.tokens_per_name
+            elif key == "content" and isinstance(value, list):
+                for submessage in value:
+                    if submessage["type"] == "text":
+                        tokens += self.calculate_text_tokens(submessage["text"])
+            elif isinstance(value, str):
+                tokens += self.calculate_text_tokens(value)
+            elif key == "content" and not isinstance(value, str):
+                raise InternalServerError(
+                    f"Unexpected type of content in message: {value!r}"
+                    f"Use MultiModalTokenizer for messages with images"
+                )
+
+        return tokens
+
+
+class MutliModalTokenizer(BaseTokenizer[MultiModalMessage]):
+    def calculate_message_tokens(
+        self, message_holder: MultiModalMessage
+    ) -> int:
+        tokens = self.tokens_per_message
+        message = message_holder.message_raw
+        # Processing text part of message
+        for key, value in message.items():
+            if key == "name":
+                tokens += self.tokens_per_name
+            elif key == "content" and isinstance(value, list):
+                for submessage in value:
+                    # Process text part of the message
+                    if submessage["type"] == "text":
+                        tokens += self.calculate_text_tokens(submessage["text"])
+            elif isinstance(value, str):
+                tokens += self.calculate_text_tokens(value)
```

Comment:
What if a submessage has `type == image`? This method **won't** throw an error.

This bug happens partially because of such a traversal of the structure, which doesn't really follow the type of this structure.

For such a simple map-reduce aggregation you typically visiting every element of a structure **once** in a top-down direction following the types - that is handling all possible cases of a value at the given node there and then.

But here the `key == "content"` is checked twice, and certain cases aren't handled **inplace** _(like type=image)_ and instead the error handling is delegated to the catch-all else guard. This leads to an error-prone code, hard to maintain and read code.

How about this:

```
for key, value in message_holder.items():
    if key == "name": collect
    elif key == "content":
        if isinstance(value, list):
            for submessage in value:
                if submessage["type"] == "text": collect
                else: fail
        elif isinstance(value, str): collect
        else: fail
```